framework/src/base/Assembly.C

```diff
@@ -1389,3 +1389,43 @@ Assembly::addJacobianOffDiagScalar(SparseMatrix<Number> & jacobian, unsigned int
       addJacobianBlock(jacobian, jacobianBlock(var_i.number(), var_j.number()), var_i.dofIndices(), var_j.dofIndices(), var_i.scalingFactor());
   }
 }
+
+void
+Assembly::cacheNodalBCJacobianEntry(numeric_index_type i, numeric_index_type j, Real value)
+{
+  _cached_nodal_bc_rows.push_back(i);
+  _cached_nodal_bc_cols.push_back(j);
+  _cached_nodal_bc_vals.push_back(value);
+}
+
+void
+Assembly::clearCachedNodalBCJacobianEntries()
+{
+  unsigned int orig_capacity = _cached_nodal_bc_rows.size();
+
+  _cached_nodal_bc_rows.clear();
+  _cached_nodal_bc_cols.clear();
+  _cached_nodal_bc_vals.clear();
+
+  // It's possible (though massively unlikely) that clear() will
+  // change the capacity of the vectors, so let's be paranoid and
+  // explicitly reserve() the same amount of memory to avoid multiple
+  // push_back() induced allocations.
+  _cached_nodal_bc_rows.reserve(orig_capacity);
```

Comment:
Wait - you can't just do 1.2*orig_capacity... or it will grow each time.  You need to keep track of the max number of entries you've seen and then set the capacity to 1.2x that.

At this point... screw it.  That's way too complicated for _tiny_ payoff...